### PR TITLE
NUTCH 2940 Develop Gradle Core Build for Apache Nutch 

### PR DESCRIPTION
**Pull request by jbsimmon**
Fixed compile-core, except still generating unneeded bin in tmp dir.

EDIT: added temp fix to prior issue (manually deleting dir). Moved forward and also fixed jar, job, and runtime.

Thanks for your contribution to [Apache Nutch](https://nutch.apache.org/)! Your help is appreciated!

Before opening the pull request, please verify that
* there is an open issue on the [Nutch issue tracker](https://issues.apache.org/jira/projects/NUTCH) which describes the problem or the improvement. We cannot accept pull requests without an issue because the change wouldn't be listed in the release notes.
* the issue ID (`NUTCH-XXXX`)
  - is referenced in the title of the pull request
  - and placed in front of your commit messages surrounded by square brackets (`[NUTCH-XXXX] Issue or pull request title`)
* commits are squashed into a single one (or few commits for larger changes)
* Java source code follows [Nutch Eclipse Code Formatting rules](https://github.com/apache/nutch/blob/master/eclipse-codeformat.xml)
* Nutch is successfully built and unit tests pass by running `ant clean runtime test`
* there should be no conflicts when merging the pull request branch into the *recent* master branch. If there are conflicts, please try to rebase the pull request branch on top of a freshly pulled master branch.

We will be able to faster integrate your pull request if these conditions are met. If you have any questions how to fix your problem or about using Nutch in general, please sign up for the [Nutch mailing list](https://nutch.apache.org/mailing_lists.html). Thanks!

### CODE REVIEW

**Comment by jbsimmon** [comment]
Most recent commit has all tasks fixed up to and including runtime, pending review

File: build.gradle.kts [hunk 1]
```diff
@@ -197,26 +197,29 @@ tasks.register<JavaCompile>("compile-core") {
...
-    options.compilerArgs.add("-Xlint-path")
+    options.compilerArgs.add("-Xlint:-path")
     options.encoding = "${project.properties["build.encoding"]}"
     options.isDebug = "${project.properties["javac.debug"]}" == "on"
     options.isDeprecation = "${project.properties["javac.deprecation"]}" == "on"
 
     copy {
-        from(layout.buildDirectory.dir("${project.properties["src.dir"]}"))
+        from("${project.properties["src.dir"]}")
         include("**/*.html")
```
Comment (lewismc) [inline build.gradle.kts]: Hi @jbsimmon these files are NOT being copied over. Please try to fix.

File: build.gradle.kts [hunk 2]
```diff
@@ -306,15 +309,15 @@ tasks.jar {
     description = "Make nutch.jar"
     dependsOn("compile-core")
 
-    from("${project.properties["conf.dir"]}/nutch-default.xml")
-    into("${project.properties["build.classes"]}")
-
-    from("${project.properties["conf.dir"]}/nutch-site.xml")
-    into("${project.properties["build.classes"]}")
+    copy {
+        from("${project.properties["conf.dir"]}/nutch-default.xml")
```
Comment (lewismc) [inline build.gradle.kts]: Neither are these files. Please fix.

File: build.gradle.kts [hunk 3]
```diff
@@ -325,34 +328,37 @@ tasks.register<Copy>("runtime")
...
+    into("${project.properties["runtime.local"]}/conf") {
+        from(layout.projectDirectory.dir("${project.properties["conf.dir"]}")) {
+            exclude("*.template")
+        }
+    }
+    into("${project.properties["runtime.local"]}/bin") {
+        from(layout.projectDirectory.dir("src/bin"))
+    }
+    into("${project.properties["runtime.local"]}/lib") {
+        from(layout.projectDirectory.dir("${project.properties["build.dir"]}/lib"))
```
Comment (lewismc) [inline build.gradle.kts]: This is not being copied either.

File: build.gradle.kts [hunk 4]
```diff
@@ -325,34 +328,37 @@ tasks.register<Copy>("runtime")
...
+    }
+    into("${project.properties["runtime.local"]}/plugins") {
+        from(layout.projectDirectory.dir("${project.properties["build.dir"]}/plugins"))
+    }
+    into("${project.properties["runtime.local"]}/test") {
```
Comment (lewismc) [inline build.gradle.kts:374]: Neither is this.

**Review by lewismc — CHANGES_REQUESTED** [review]
Also note that the `compile-core-test` task is not working.

**Comment by lewismc** [comment]
Also note `gradle clean jar` results in the following issue. Please take a look into it.
```
Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.3.3/userguide/command_line_interface.html#sec:command_line_warnings

Execution optimizations have been disabled for 1 invalid unit(s) of work during this build to ensure correctness.
Please consult deprecation warnings for more details.

BUILD SUCCESSFUL in 4s
9 actionable tasks: 4 executed, 5 up-to-date
```
Thank you @jbsimmon

**Comment by jbsimmon** [comment]
@lewismc Fixed errors with files not copying over, and also fixed the deprecated gradle feature error. Tried to implement compile-core-test but still getting an error about non-existent packages, not sure if dependency related. Please take a look at most recent commit for changes, thank you!

File: build.gradle.kts [hunk 5]
```diff
@@ -188,35 +189,44 @@ tasks.register<Sync>("resolve-default") {
...
+
 tasks.register("compile") {
     description = "Compile all Java files"
     dependsOn("compile-core","compile-plugins")
 }
 
 tasks.register<JavaCompile>("compile-core") {
     description = "Compile core Java files only"
-    dependsOn("init-nutch","resolve-default")
+    dependsOn("init-nutch","resolve-default","compileJava")
```
Comment (lewismc) [inline build.gradle.kts:207]: Is this necessary? What does `compileJava` do?

File: build.gradle.kts [hunk 6]
```diff
@@ -188,35 +189,44 @@ tasks.register<Sync>("resolve-default") {
...
-
+    
     copy {
-        from(layout.buildDirectory.dir("${project.properties["src.dir"]}"))
-        include("**/*.html")
-        include("**/*.css")
-        include("**/*.properties")
-        into(layout.buildDirectory.dir("${project.properties["build.classes"]}"))
+        from(layout.projectDirectory.dir("${project.properties["src.dir"]}"))
+        include("**/*.html", "**/*.css", "**/*.properties")
```
Comment (lewismc) [inline build.gradle.kts:225]: Please remove this line. [overview.html](https://github.com/csci401-apache-nutch/nutch/blob/NUTCH-2940/src/java/overview.html) is still not being copied over to `build/classes/overview.html`

File: build.gradle.kts [hunk 7]
```diff
@@ -306,15 +316,19 @@ tasks.jar {
...
     dependsOn("compile-core")
 
-    from("${project.properties["conf.dir"]}/nutch-default.xml")
-    into("${project.properties["build.classes"]}")
-
-    from("${project.properties["conf.dir"]}/nutch-site.xml")
-    into("${project.properties["build.classes"]}")
+    copy {
+        from(
+            file(layout.projectDirectory.dir("${project.properties["conf.dir"]}/nutch-default.xml")),
```
Comment (lewismc) [inline build.gradle.kts:321]: These files are also not being copied over. Please verify that they are.

File: build.gradle.kts [hunk 8]
```diff
@@ -306,15 +316,19 @@ tasks.jar {
...
+            file(layout.projectDirectory.dir("${project.properties["conf.dir"]}/nutch-site.xml"))
+        )
+        // from("${project.properties["conf.dir"]}/nutch-default.xml")
```
Comment (lewismc) [inline build.gradle.kts:324]: If not being used then please remove.

File: build.gradle.kts [hunk 9]
```diff
@@ -325,34 +339,41 @@ tasks.register<Copy>("runtime")
...
-    from(layout.buildDirectory.dir("${project.properties["runtime.local"]}/conf")) {
-        exclude("*.template")
+    into("${project.properties["runtime.deploy"]}") {
+        from(layout.projectDirectory.dir("${project.properties["build.dir"]}/${project.properties["final.name"]}.job"))
+    }
+    into("${project.properties["runtime.deploy"]}/bin") {
+        from(layout.projectDirectory.dir("src/bin"))
+    }
+    into("${project.properties["runtime.local"]}/lib") {
+        // from(
```
Comment (lewismc) [inline build.gradle.kts:351]: Why is this copied out? We are trying to replicate the behaviour from the `ant` build. This directory needs to be present and on the classpath in order for Nutch to be able to run.

**Review by lewismc — CHANGES_REQUESTED** [review]
Hi @jbsimmon thanks for attempting to update more. Please see my comments. 
In particular, please pay attention to the files being copied over. Namely, `nutch-site` and `nutch-default.xml`. 

An interesting observation, if you look at the size of the generated `.job` artifact you will see that it is small `912kb` in comparison to the `282MB` when built with `ant. Maybe this just has to do with all the plugin being packaged into the artifact.

File: build.gradle.kts (hunk 5 above)
Comment (jbsimmon) [reply build.gradle.kts:207]: This dependency is necessary to fix the issue you noted that previously came up when running gradle clean jar. I am not sure if there is an alternate way to fix it, but this explicit declaration seems to work in this case.

**Comment by lewismc** [comment]
Also, when I run `job` and look into the artifact I see the following
```
jar tf build/apache-nutch-1.19-SNAPSHOT.job
META-INF/
META-INF/MANIFEST.MF
classes/
classes/plugins
org/
org/apache/
org/apache/nutch/
org/apache/nutch/tools/
org/apache/nutch/exchange/
org/apache/nutch/metadata/
org/apache/nutch/fetcher/
org/apache/nutch/tools/warc/
org/apache/nutch/crawl/
org/apache/nutch/util/
org/apache/nutch/segment/
org/apache/nutch/service/
org/apache/nutch/service/model/
org/apache/nutch/service/model/response/
org/apache/nutch/tools/arc/
org/apache/nutch/net/
org/apache/nutch/hostdb/
org/apache/nutch/protocol/
org/apache/nutch/service/resources/
org/apache/nutch/scoring/
org/apache/nutch/scoring/webgraph/
org/apache/nutch/net/protocols/
org/apache/nutch/plugin/
org/apache/nutch/util/domain/
org/apache/nutch/indexer/
org/apache/nutch/publisher/
org/apache/nutch/parse/
org/apache/nutch/service/model/request/
org/apache/nutch/service/impl/
```

No compiled classes, configuration resources or libraries are being packaged into `.job`. This needs to be fixed.

**Comment by lewismc** [comment]
Going to merge @jbsimmon we can open new PR's when we identify issues.

**Pull request merged**